### PR TITLE
runtime: add reconnect timeout for vhost user block

### PR DESCRIPTION
**Pull request by zhaojizhuang**
Fixes: #6075 

add reconnect timeout for vhost user block

allow users set timeout via `configurationg.toml` or pod `annotations`:

1. `configurationg.toml` 
```shell
#configuration-qemu.toml
vhost_user_block_reconnect_timeout_sec = 1
```
2. pod `annotations`

```yaml
annotations:
  io.katacontainers.config.hypervisor. vhost_user_block_reconnect_timeout_sec: "1"
```

Signed-off-by: zhaojizhuang <571130360@qq.com>

### CODE REVIEW

**Comment by zhaojizhuang** [comment]
cc @liubin

File: src/runtime/pkg/govmm/qemu/qmp.go [hunk 1]
```diff
@@ -1512,9 +1512,9 @@ func (q *QMP) ExecuteGetFD(ctx context.Context, fdname string, fd *os.File) erro
 }
 
 // ExecuteCharDevUnixSocketAdd adds a character device using as backend a unix socket,
-// id is an identifier for the device, path specifies the local path of the unix socket,
+// id is an identifier for the device, pa：th specifies the local path of the unix socket,
```
Comment (YushuoEdge) [inline src/runtime/pkg/govmm/qemu/qmp.go:1515]: typo?

File: src/runtime/pkg/device/config/config.go [hunk 2]
```diff
@@ -97,6 +97,13 @@ const (
 	VhostUserSCSIMajor = 242
 )
 
+const (
+	//The timeout for reconnecting on non-server sockets when the remote end
+	//goes away.  qemu will delay this many seconds and then attempt to reconnect.  Zero
+	//disables reconnecting, and is the default.
+	VhostUserBlkReconnectTimeOut = 1
```
Comment (fengwang666) [inline src/runtime/pkg/device/config/config.go:104]: How did you pick 1 second?

File: src/runtime/pkg/device/config/config.go [hunk 3]
```diff
@@ -320,6 +327,9 @@ type VhostUserDeviceAttrs struct {
 	CacheSize uint32
 
 	QueueSize uint32
+
+	// Reconnect timeout for socket of vhost user block device
+	ReconnectTime int
```
Comment (fengwang666) [inline src/runtime/pkg/device/config/config.go:332]: Do we want to expose this as a config in the configuration.toml?

**Review by fengwang666 — COMMENTED** [review]
Thanks! One question, will it keep retrying indefinitely if spdk is not up?

File: src/runtime/pkg/govmm/qemu/qmp.go (hunk 1 above)
Comment (zhaojizhuang) [reply src/runtime/pkg/govmm/qemu/qmp.go:1515]: sorry, typo

File: src/runtime/pkg/device/config/config.go (hunk 3 above)
Comment (zhaojizhuang) [reply src/runtime/pkg/device/config/config.go:332]: Yes, If we expose this as a config will be better

File: src/runtime/pkg/device/config/config.go (hunk 2 above)
Comment (zhaojizhuang) [reply src/runtime/pkg/device/config/config.go:104]: Just a default setting

**Comment by zhaojizhuang** [comment]
> Thanks! One question, will it keep retrying indefinitely if spdk is not up?

@fengwang666 Yeah, qemu will reconnect spdk after every `reconnect` seconds as a client, until spdk up

**Comment by zhaojizhuang** [comment]
> Fixes: #6075
> 
> add reconnect timeout for vhost user block
> 
> allow users set timeout via `configurationg.toml` or pod `annotations`:
> 
> 1. `configurationg.toml`
> 
> ```shell
> #configuration-qemu.toml
> vhost_user_block_device_reconnect = 1
> ```
> 
> 2. pod `annotations`
> 
> ```yaml
> annotations:
>   io.katacontainers.config.hypervisor.vhost_user_block_device_reconnect: "1"
> ```
> 
> Signed-off-by: zhaojizhuang [571130360@qq.com](mailto:571130360@qq.com)

updated,  expose configuration via configuration.toml, cc @fengwang666

File: src/runtime/pkg/govmm/qemu/qmp.go (hunk 1 above)
Comment (zhaojizhuang) [reply src/runtime/pkg/govmm/qemu/qmp.go]: done

**Comment by zhaojizhuang** [comment]
ping @fengwang666 @liubin

File: src/runtime/pkg/device/drivers/vhost_user_blk.go [hunk 4]
```diff
@@ -93,6 +96,24 @@ func (device *VhostUserBlkDevice) Attach(ctx context.Context, devReceiver api.De
...
 
+func vhostUserReconnect(customOptions map[string]string) uint32 {
+	var vhostUserBlkReconnectOpt uint32
+
+	if customOptions == nil {
+		vhostUserBlkReconnectOpt = config.DefaultVhostUserBlkReconnectTimeOut
+	} else {
+		reconnectStr := customOptions[config.VhostUserBlkReconnectTimeOutOpt]
+		if reconnect, err := strconv.Atoi(reconnectStr); err != nil {
+			vhostUserBlkReconnectOpt = 0
```
Comment (bergwolf) [inline src/runtime/pkg/device/drivers/vhost_user_blk.go:107]: use the default timeout instead of 0?

**Review by bergwolf — COMMENTED** [review]
Generally LGTM, just two minor comments inline. Thanks!

File: src/runtime/config/configuration-qemu.toml.in [hunk 5]
```diff
@@ -295,6 +295,11 @@ vhost_user_store_path = "@DEFVHOSTUSERSTOREPATH@"
 # Your distribution recommends: @DEFVALIDVHOSTUSERSTOREPATHS@
 valid_vhost_user_store_paths = @DEFVALIDVHOSTUSERSTOREPATHS@
 
+# The timeout for reconnecting on non-server spdk sockets when the remote end goes away.
+# qemu will delay this many seconds and then attempt to reconnect.
+# Zero disables reconnecting, and the default is 1 seconds.
+vhost_user_block_device_reconnect = 1
```
[the remark this answers is not in the document]
Comment (zhaojizhuang) [reply src/runtime/config/configuration-qemu.toml.in:301]: I have no objection

File: src/runtime/pkg/device/drivers/vhost_user_blk.go (hunk 4 above)
Comment (zhaojizhuang) [reply src/runtime/pkg/device/drivers/vhost_user_blk.go]: done

**Comment by zhaojizhuang** [comment]
ping @bergwolf @liubin

File: docs/how-to/how-to-set-sandbox-config-kata.md [hunk 6]
```diff
@@ -57,6 +57,7 @@ There are several kinds of Kata configurations and they are listed below.
 | `io.katacontainers.config.hypervisor.enable_iothreads` | `boolean`| enable IO to be processed in a separate thread. Supported currently for virtio-`scsi` driver |
 | `io.katacontainers.config.hypervisor.enable_mem_prealloc` | `boolean` | the memory space used for `nvdimm` device by the hypervisor |
 | `io.katacontainers.config.hypervisor.enable_vhost_user_store` | `boolean` | enable vhost-user storage device (QEMU) |
+| `io.katacontainers.config.hypervisor.vhost_user_block_reconnect_timeout_sec` | `string`| the timeout for reconnecting SPDK (QEMU)
```
Comment (liubin) [inline docs/how-to/how-to-set-sandbox-config-kata.md:60]: Is it reasonable to remove the `block` from the annotation/var name that let it be for all vhost user devices, though only block device is supported now?

File: src/runtime/pkg/device/config/config.go [hunk 7]
```diff
@@ -320,6 +331,9 @@ type VhostUserDeviceAttrs struct {
 	CacheSize uint32
 
 	QueueSize uint32
+
+	// Reconnect timeout for socket of vhost user block device
+	ReconnectTime uint32
```
Comment (liubin) [inline src/runtime/pkg/device/config/config.go:336]: Nit: change to `ReconnectTimeout`.

And suggest adding a `timeout` suffix to other var names to let readers know that it is a timeout but not a reconnect count.

File: src/runtime/pkg/device/drivers/vhost_user_blk.go [hunk 8]
```diff
@@ -93,6 +96,24 @@ func (device *VhostUserBlkDevice) Attach(ctx context.Context, devReceiver api.De
 	return nil
 }
 
+func vhostUserReconnect(customOptions map[string]string) uint32 {
+	var vhostUserBlkReconnectOpt uint32
```
Comment (liubin) [inline src/runtime/pkg/device/drivers/vhost_user_blk.go:100]: Nit: vhostUserBlkReconnectOpt -> vhostUserBlkReconnectTimeout

File: docs/how-to/how-to-set-sandbox-config-kata.md (hunk 6 above)
Comment (zhaojizhuang) [reply docs/how-to/how-to-set-sandbox-config-kata.md:60]: sounds like a good idea

File: src/runtime/pkg/device/config/config.go (hunk 7 above)
Comment (zhaojizhuang) [reply src/runtime/pkg/device/config/config.go:336]: ok

**Comment by zhaojizhuang** [comment]
updated @liubin

**Comment by liubin** [comment]
@YushuoEdge @bergwolf @fengwang666 PTAL

**Review by bergwolf — APPROVED** [review]
lgtm!

File: src/runtime/pkg/device/config/config.go [hunk 9]
```diff
@@ -87,6 +87,8 @@ const (
 	// Define the string key for DriverOptions in DeviceInfo struct
 	FsTypeOpt      = "fstype"
 	BlockDriverOpt = "block-driver"
+
+	VhostUserBlkReconnectTimeOutOpt = "vhost-user-blk-reconnect"
```
Comment (liubin) [inline src/runtime/pkg/device/config/config.go:91]: Should here remove `Blk`?

File: src/runtime/pkg/device/config/config.go [hunk 10]
```diff
@@ -97,6 +99,15 @@ const (
 	VhostUserSCSIMajor = 242
 )
 
+const (
+
+	// The timeout for reconnecting on non-server sockets when the remote end
+	// goes away.
+	// qemu will delay this many seconds and then attempt to reconnect.  Zero
+	// disables reconnecting, and is the default.
+	DefaultVhostUserBlkReconnectTimeOut = 0
```
Comment (liubin) [inline src/runtime/pkg/device/config/config.go:108]: Ditto.

File: src/runtime/pkg/device/drivers/vhost_user_blk.go [hunk 11]
```diff
@@ -93,6 +96,24 @@ func (device *VhostUserBlkDevice) Attach(ctx context.Context, devReceiver api.De
 	return nil
 }
 
+func vhostUserReconnect(customOptions map[string]string) uint32 {
+	var vhostUserBlkReconnectTimeout uint32
```
Comment (liubin) [inline src/runtime/pkg/device/drivers/vhost_user_blk.go:100]: Ditto.

File: src/runtime/pkg/device/manager/manager.go [hunk 12]
```diff
@@ -42,17 +43,20 @@ type deviceManager struct {
 	sync.RWMutex
 
 	vhostUserStoreEnabled bool
+
+	vhostUserBlkReconnect uint32
```
Comment (liubin) [inline src/runtime/pkg/device/manager/manager.go:47]: Ditto.

File: src/runtime/pkg/device/manager/manager.go [hunk 13]
```diff
@@ -42,17 +43,20 @@ type deviceManager struct {
...
 }
 
 func deviceLogger() *logrus.Entry {
 	return api.DeviceLogger().WithField("subsystem", "deviceManager")
 }
 
 // NewDeviceManager creates a deviceManager object behaved as api.DeviceManager
-func NewDeviceManager(blockDriver string, vhostUserStoreEnabled bool, vhostUserStorePath string, devices []api.Device) api.DeviceManager {
+func NewDeviceManager(blockDriver string, vhostUserStoreEnabled bool, vhostUserStorePath string, vhostUserBlkReconnect uint32, devices []api.Device) api.DeviceManager {
```
Comment (liubin) [inline src/runtime/pkg/device/manager/manager.go:55]: Ditto.

File: src/runtime/pkg/device/config/config.go (hunk 9 above)
Comment (zhaojizhuang) [reply src/runtime/pkg/device/config/config.go]: of course we can

**Review by liubin — APPROVED** [review]

**Comment by zhaojizhuang** [comment]
CI  failed, anyone fix it ?

**Comment by liubin** [comment]
@zhaojizhuang file conflicting, you should rebase.

**Comment by zhaojizhuang** [comment]
updated，rebase and fix conflicting

**Pull request merged**